orc8r/cloud/go/service/middleware/unary/identity_decorator.go

```diff
@@ -112,70 +112,97 @@ func SetIdentityFromContext(ctx context.Context, _ interface{}, info *grpc.Unary
 		return newCtx, newReq, resp, err
 	}
 
+	// check for allowlisted calls only once, since info does not change
+	// Bypasses for allowlisted methods are given further down, when the
+	// call would be rejected otherwise
+	allow := isAllowlistedCall(info)
+
 	// First, try to find the caller's identity
 	snlist, snok := ctxMetadata[CLIENT_CERT_SN_KEY]
-	if snok { // there is a certificate serial number (CSN) list in CTX
-		if len(snlist) != 1 {
-			// there can be only one CSN, error out if not
-			glog.Infof("Multiple CSNs found in metadata: %+v", ctxMetadata)
+	if snok && len(snlist) != 1 {
+		// 3.
+		// there is a certificate serial number (CSN) list in CTX
+		// there can be only one CSN, error out if not
+		glog.Infof("Multiple CSNs found in metadata: %+v", ctxMetadata)
+		if !allow {
 			err = status.Error(codes.Unauthenticated, "Multiple CSNs present")
-		} else {
-			// One CSN is found, find Identity associated with it
-			var gwIdentity *protos.Identity
-			var certExpTime int64
-			// Check if SN is the reserved value used for all inter-orc8r calls
-			if snlist[0] == unarylib.ORC8R_CLIENT_CERT_VALUE {
-				return newCtx, newReq, resp, nil
-			}
-			gwIdentity, certExpTime, err = findGatewayIdentity(ctx, snlist[0], ctxMetadata)
-			if err == nil {
-				// If a valid GW Identity is found, add it into CTX for use
-				// by the callee
-				newCtx = gwIdentity.NewContextWithIdentity(protos.NewContextWithCertExpiration(ctx, certExpTime))
-				return newCtx, newReq, resp, err
-			}
 		}
-	} else {
-		// No CSNs found, check there is also no Certificate Common Names
-		if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
-			// CN header is present while SN header is missing - possible
-			// security hack, either both or neither of the headers should be
-			// set
-			glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
+	} else if snok {
```

Comment:
For the most common case it should now be the same efficiency, and for the case that there are multiple CSNs for the same key, it is just an if check of a bool more. I'd argue that having all distinct cases on the same level might be more readable, since there are quite a few cases (might be subjective)